### ISSUE TITLE
High CPU usage with conky conky-1.11.4 and conky-1.11.5 in virtualbox

### ISSUE THREAD

**Issue by tiiiecherle** [brndnmtthws/conky#876]
[conky_archvbox.txt](https://github.com/brndnmtthws/conky/files/3501268/conky_archvbox.txt)

Hey, 

I'm using archlinux with all packages up-to-date.

Starting with conky 1.11.4 the CPU usage rises from about 2% when doing nothing after login to around 10%. Killing conky brings the cpu back to normal. Downgrading to 1.11.3 solves the problem, too.

On my "real" system the cpu usage is higher, too. But not that much as in the virtualbox.

I attached my conky config.

Thanks for looking into it in advance.

**Comment by cspack77** [comment]
I'm seeing the same thing (not using a VM). CPU usage increases starting with v1.11.4. Git bisect points to this commit: [Add support for nested |exec| expressions](https://github.com/brndnmtthws/conky/commit/c352069b31b2dbb4d87503603d6c6d3ddd517df6) (I'm not using this feature though).

In my testing (v1.11.5) the CPU increase is especially high if I have an execpi command in my config. Also with execpi I noticed the interval specified seems to be cut in half. So if I specify execpi 60, the command executes every 30 seconds. Additionally, if I have any execi commands they also have their interval cut in half. If I remove the execpi, any execi commands execute at the correct interval.

**Comment by CareF** [comment]
I confirm high CPU usage when there's execpi. A minimum config file with top monitor is: 

```
conky.config = {
    own_window = true,
}

conky.text = [[
${execpi 10 echo "test"}
${color grey}CPU usage          PID    CPU%   MEM%
${color #ddaa00} ${top name 1} ${top pid 1} ${top cpu 1} ${top mem 1}
$color ${top name 2} ${top pid 2} ${top cpu 2} ${top mem 2}
$color ${top name 3} ${top pid 3} ${top cpu 3} ${top mem 3}
$color ${top name 4} ${top pid 4} ${top cpu 4} ${top mem 4}
$color ${top name 5} ${top pid 5} ${top cpu 5} ${top mem 6}
]]
```

On my laptop often times conky goes to 10% CPU, while if I change `execpi` to `execi` there's not such issue.

**Comment by rtoijala** [comment]
I have bisected this to:
```
commit c352069b31b2dbb4d87503603d6c6d3ddd517df6 (HEAD, refs/bisect/bad)
Author: Kushagra Sinha <sinhak@google.com>
Date:   Sun Mar 3 17:45:02 2019 +0100

    Add support for nested |exec| expressions
    
    Nested exec expressions like ${execp echo '${execp echo hi}'} are not
    evaluated correctly because parsing expressions like these generates and
    registers new callbacks but we never wait for these callbacks to
    complete before returning the result of the evaluation.
    
    Fix this by re-evaluating expressions after running all pending
    callbacks.
```
It adds to `evaluate` an extra call to `parse_conky_vars`  and to `run_all_callbacks`. The code has changed a bit since then, but commenting out the `run_all_callbacks` line brings the CPU usage back from ~20% to around 2%, where it was before the commit.

**Comment by tiiiecherle** [comment]
Thanks a lot ;)

**Comment by stevenxxiu** [comment]
Appears to be fixed by https://github.com/brndnmtthws/conky/pull/1093. My CPU usage is from 1% to 0.25% now.

**Comment by brndnmtthws** [comment]
> Appears to be fixed by #1093. My CPU usage is from 1% to 0.25% now.

Sweet! Thanks for the update. Going to close this now.
